### PR TITLE
Ngc 722 fft unzip update

### PR DESCRIPTION
**Pull request by andrewvanderbyl**
Made some changes to documentation describing FFT optimisation.

### CODE REVIEW

File: doc/fgpu.design.rst [hunk 1]
```diff
@@ -177,51 +177,129 @@ slow and require multiple passes over the memory, because
...
+
+Hermitian symmetry means that :math:`U'_k = U_{-k} = \overline{U_k}` where
+the 'overline' in :math:`\overline{U_k}` denotes conjugation. This is effectively saying that by taking the
+reverse indices in :math:`U_k` we get a conjugated result.
+
+Why is this so? Going back to the original definition for the DFT we saw the complex exponential
+:math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the
+frequency component under computation for the input sequence :math:`x_i`. If :math:`k` is reversed
+(i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2n}\cdot i(k)}` as the negaitive
+in :math:`-k` multiplies out.
```
Comment (bmerry) [inline doc/fgpu.design.rst:238]: @james-smith-za do you have an opinion on whether this proof of a standard Fourier result in Fourier is adding anything? I feel like it interrupts the flow of the discussion.

File: doc/fgpu.design.rst [hunk 2]
```diff
@@ -177,51 +177,129 @@ slow and require multiple passes over the memory, because
...
+
+Moving back to :math:`U` and :math:`V` components then :math:`U' = \overline{U}` and similarly :math:`V' = \overline{V}`.
+Going full circle (for completeness) if one was to conjugate the reversal (i.e. :math:`\overline{U'}`) then since
+:math:`U' = \overline{U}`, :math:`\overline{U'} = \overline{\overline{U}} = U`. Likewise for
+:math:`\overline{V'} = \overline{\overline{V}} = V`.
```
Comment (bmerry) [inline doc/fgpu.design.rst:243]: @james-smith-za again I think we need a tie-breaker from you - I think stating that conjugating twice cancels out doesn't really add anything.

File: doc/fgpu.design.rst [hunk 3]
```diff
@@ -177,51 +177,129 @@ slow and require multiple passes over the memory, because
...
+
+Why is this important? Previously we stated that :math:`Z = U + jV`. Now we can consider the reverse (which becomes
+the conjugate as described above) of :math:`Z`, namely :math:`Z'`.
```
Comment (bmerry) [inline doc/fgpu.design.rst:246]:
```suggestion
Why is this important? Previously we stated that :math:`Z = U + jV`. Now we can consider the reverse
of :math:`Z`, namely :math:`Z'`.
```

Z' is **not** the conjugate of Z.

**Review by bmerry — CHANGES_REQUESTED** [review]

File: doc/fgpu.design.rst [hunk 4]
```diff
@@ -233,17 +311,121 @@ into the postprocessing kernel (see the next section).
...
+
+.. image:: images/z_array_breakdown.png
+   :width: 600
+
+Right, so we have separate sub-arrays as indexed form the initial array, what happens next? These various
+:math:`z^{r}` arrays are fed to cuFFT yielding 4 complex-to-complex transforms. These separate transforms
+now need to be combined to form a single real-to-complex transform of the full initial size.
+
+To see how the :math:`k` indexing works, :math:`k = pm + s` and is dealt with in a similar manner as above.
+The parameter :math:`m=8192` is static (in this case), and :math:`p` has a range :math:`0` to :math:`n-1`
```
Comment (bmerry) [inline doc/fgpu.design.rst:382]:
```suggestion
The parameter :math:`m=8192` is static (in this case), and :math:`p` has a range :math:`0` to :math:`n-1`
```

Not sure what you mean by "static".

**Review by bmerry — CHANGES_REQUESTED** [review]

File: doc/fgpu.design.rst [hunk 5]
```diff
@@ -233,17 +311,121 @@ into the postprocessing kernel (see the next section).
 
 Unzipping the FFT
 ~~~~~~~~~~~~~~~~~
-From here we'll assume all transforms are complex-to-complex unless specified
-otherwise. The Cooley-Tukey algorithm allows a transform of size :math:`c =
-mn` to be decomposed into :math:`n` transforms of size :math:`m` followed by
-:math:`m` transforms of size :math:`n`. We'll refer to :math:`n` as the
+Right - lets get practical and show how we *actually* implement this. From here we'll assume all
```
[the remark this answers is not in the document]
Comment (bmerry) [reply doc/fgpu.design.rst:314]: Try the suggested change I made and check that it looks sensible in the output.

**Comment by andrewvanderbyl** [comment]
Updated made. Another eyeballing requested :)

File: doc/fgpu.design.rst [hunk 6]
```diff
@@ -177,51 +177,132 @@ slow and require multiple passes over the memory, because
...
-:math:`X_k` be the Fourier transform (we divide by :math:`2n` rather than
-:math:`n` because that's the number of input samples used).
+Now for some notation to see how this works. We start by treating :math:`x` (a real array of length N) as if it is
+a complex array :math:`z` of length N, with each adjacent pair of real values in :math:`x` interpreted as the
+real and imaginary components of a complex value, and computing the Fourier transform of :math:`z`. Formally,
+let :math:`u_i = x_{2i}` and :math:`v_i = x_{2i+1}`. Then :math:`z_i = u_i + jv_i = x_{2i} + j x_{2i+1}`.
+
+For clarity, array :math:`x` has length :math:`2N` but is divided into even- and odd- samples
+each of length :math:`N`. In other words, each adjacent pair of real values in :math:`x` is interpreted as the real
+and imaginary components of a complex value.
```
Comment (bmerry) [inline doc/fgpu.design.rst:217]: Can delete this paragraph - the previous one now says all this.

File: doc/fgpu.design.rst [hunk 7]
```diff
@@ -177,51 +177,132 @@ slow and require multiple passes over the memory, because
...
+is intentionally kept small for efficiency reasons and later combined (same process as the FFT) to form
+the larger transform size. This is a multi-step process and requires some extra notation and math tricks.
 
 Real-to-complex transform
 ~~~~~~~~~~~~~~~~~~~~~~~~~
-Let's introduce some notation to see how this works. Let :math:`n` be the
-number of channels, :math:`x_i` be the (real) time-domain samples and
-:math:`X_k` be the Fourier transform (we divide by :math:`2n` rather than
-:math:`n` because that's the number of input samples used).
+Now for some notation to see how this works. We start by treating :math:`x` (a real array of length N) as if it is
```
Comment (bmerry) [inline doc/fgpu.design.rst:210]:
```suggestion
Now for some notation to see how this works. We start by treating :math:`x` (a real array of length 2N) as if it is
```

File: doc/fgpu.design.rst [hunk 8]
```diff
@@ -177,51 +177,132 @@ slow and require multiple passes over the memory, because
...
+let :math:`u_i = x_{2i}` and :math:`v_i = x_{2i+1}`. Then :math:`z_i = u_i + jv_i = x_{2i} + j x_{2i+1}`.
+
+For clarity, array :math:`x` has length :math:`2N` but is divided into even- and odd- samples
+each of length :math:`N`. In other words, each adjacent pair of real values in :math:`x` is interpreted as the real
+and imaginary components of a complex value.
+
+We will start by computing the Fourier transform of :math:`z`. This is convenient to do because :math:`z` has the
+identical memory layout to :math:`x`, so we can obtain it just by viewing the same memory as containing N complex
+numbers instead of 2N real numbers. In other words, each adjacent pair of real values in :math:`x` is interpreted
+as the real and imaginary components of a complex value.
```
Comment (bmerry) [inline doc/fgpu.design.rst:222]: And you can delete everything after the first sentence in this paragraph too (as well as the paragraph break at the end).

File: doc/fgpu.design.rst [hunk 9]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+:math:`8191 \cdot 4 + 3`   32767
+========================  ========
+
+What this shows is that each sub-array consists of samples from the initial array :math:`z` indexed
+by :math:`i=qn+r` where each sample is every :math:`4^{th}` and offset by :math:`r`. Pictorially
+this looks like,
+
+.. image:: images/z_array.png
+   :width: 600
+(click image to enlarge)
```
Comment (bmerry) [inline doc/fgpu.design.rst:399]: Sphinx is giving a warning:
```
katgpucbf/doc/fgpu.design.rst:399: WARNING: Explicit markup ends without a blank line; unexpected unindent.
```

It's also not going to make any sense in the PDF output (I think Sphinx has some directives that let you limit some stuff to a particular output format).

I feel like just using a bigger font (and maybe fewer elements) would make it readable without the need to enlarge.

File: doc/fgpu.design.rst [hunk 10]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+...               ...
+...               ...
+:math:`8191 \cdot 4 + 3`   32767
+========================  ========
+
+What this shows is that each sub-array consists of samples from the initial array :math:`z` indexed
+by :math:`i=qn+r` where each sample is every :math:`4^{th}` and offset by :math:`r`. Pictorially
+this looks like,
+
+.. image:: images/z_array.png
```
Comment (bmerry) [inline doc/fgpu.design.rst:397]: Where is the source for this image?

File: doc/fgpu.design.rst [hunk 11]
```diff
@@ -177,51 +177,132 @@ slow and require multiple passes over the memory, because
...
-post-processing kernel (using the Cooley-Tukey algorithm), and also handle
-fixing up the real-to-complex transformation.
+post-processing kernel, and also handle fixing up the real-to-complex transformation.
+This is achieved by decomposing the transformation into separately-computed smaller parts
+(using the Cooley-Tukey algorithm). Part of the Fourier transform is computed using
+cuFFT and the final stage (post-processing kernel) of the process includes one round of
+Cooley-Tukey computation and the computation to form the real-to-complex transformation.
+
+To start, let's consider the traditional equation for the Fourier Transform. Let :math:`N`
+be the number of channels into which we wish to decompose the input sequence into, and let
```
Comment (bmerry) [inline doc/fgpu.design.rst:190]:
```suggestion
be the number of channels into which we wish to decompose the input sequence, and let
```

File: doc/fgpu.design.rst [hunk 12]
```diff
@@ -177,51 +177,132 @@ slow and require multiple passes over the memory, because
...
+:math:`x_i` be the (real) time-domain samples (:math:`0 \le i < 2N`) and :math:`X_k` be its discrete
+Fourier transform (DFT). Because the time domain is real, the frequency domain is Hermitian symmetric,
+and we only need to compute half of it to recover all the information. We thus only need to consider
+:math:`k` from :math:`0` to :math:`N-1` (this loses information about :math:`X_N`, but it is convenient
+to discard it and thus have a power-of-two number of outputs).
+
+.. math:: X_k = \sum_{i=0}^{2N-1} e^{\frac{-2\pi j}{2N}\cdot ik} x_i.
+
+We know that a direct implementation of the DFT is inefficient and alternative, more efficient
+means exist to perform this computation. One such method is the FFT introduce by Cooley-Tukey
```
Comment (bmerry) [inline doc/fgpu.design.rst:200]:
```suggestion
means exist to perform this computation. One such method is the FFT introduced by Cooley-Tukey
```

File: doc/fgpu.design.rst [hunk 13]
```diff
@@ -233,17 +314,16 @@ into the postprocessing kernel (see the next section).
...
 "unzipping factor". We will keep it small (typically not more than 4), as the
-implementation requires registers proportional to this factor.
+implementation requires registers proportional to this factor. We are now going to go step-by-step and
+separate the input array :math:`x` into :math:`n` parts of size :math:`m` with each part operated on using a
+Fourier transform.
 
-To recap the Cooley-Tukey algorithm: let a time-domain index :math:`i` be
-written as :math:`qn + r` and a frequency-domain index :math:`k` be
-written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
-\dots, z_{(m-1)n+r}`, and denote its Fourier transform by :math:`Z^r`. Then
```
Comment (bmerry) [inline doc/fgpu.design.rst:246]: By moving these definitions to later on, you're now using them before they're defined.

File: doc/fgpu.design.rst [hunk 14]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
-and over multiple spectra at the same time as it only supports a single batch
-dimension with corresponding stride. We solve this by modifying the PFB kernel
-to reorder its output such that each :math:`z^r` is output contiguously. This
-can be done by shuffling some bits in the output index (because we assume
-powers of two everywhere).
+Now let's walk through each part. To recap the indexing used in the Cooley-Tukey algorithm: let a
+time-domain index :math:`i` be written as :math:`qn + r` and a frequency-domain index :math:`k` be
+written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r}, \dots, z_{(m-1)n+r}`,
+and denote its Fourier transform by :math:`Z^r`. It is worthwhile to point out that the superscript
+:math:`r` *does not* denote raise to the power but rather is a means to indicate an :math:`r^{th}` array.
```
Comment (bmerry) [inline doc/fgpu.design.rst:346]:
```suggestion
:math:`r` *does not* denote exponentiation but rather is a means to indicate an :math:`r^{th}` array.
```

File: doc/fgpu.design.rst [hunk 15]
```diff
@@ -233,17 +314,16 @@ into the postprocessing kernel (see the next section).
...
-mn` to be decomposed into :math:`n` transforms of size :math:`m` followed by
-:math:`m` transforms of size :math:`n`. We'll refer to :math:`n` as the
+Right --- lets get practical and show how we *actually* implement this. From here we'll assume all
+transforms are complex-to-complex unless specified otherwise. Firstly, some recap: the Cooley-Tukey
+algorithm allows a transform of size :math:`N = mn` to be decomposed into :math:`n` transforms of
+size :math:`m` followed by :math:`m` transforms of size :math:`n`. We'll refer to :math:`n` as the
 "unzipping factor". We will keep it small (typically not more than 4), as the
-implementation requires registers proportional to this factor.
+implementation requires registers proportional to this factor. We are now going to go step-by-step and
+separate the input array :math:`x` into :math:`n` parts of size :math:`m` with each part operated on using a
```
Comment (bmerry) [inline doc/fgpu.design.rst:323]:
```suggestion
separate the input array :math:`z` into :math:`n` parts of size :math:`m` with each part operated on using a
```

File: doc/fgpu.design.rst [hunk 16]
```diff
@@ -233,17 +311,121 @@ into the postprocessing kernel (see the next section).
...
-written as :math:`qn + r` and a frequency-domain index :math:`k` be
-written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
-\dots, z_{(m-1)n+r}`, and denote its Fourier transform by :math:`Z^r`. Then
+To recap the Cooley-Tukey algorithm: let a time-domain index :math:`i` be written as :math:`qn + r`
+and a frequency-domain index :math:`k` be written as :math:`pm + s`. Let :math:`z^r` denote the array
+:math:`z_r, z_{n+r}, \dots, z_{(m-1)n+r}`, and denote its Fourier transform by :math:`Z^r`. It is
+worthwhile to point out that the superscript :math:`r` *does not* denote raise to the power but rather
+is a means to indicate an :math:`r^{th}` array. In practice this :math:`r^{th}` array is a subset (part)
+of the larger :math:`z` array of input data. Let's unpack this a bit further - What is actually happening
+is that the initial array :math:`z` is divided into :math:`n=4` separate arrays each of
```
[the remark this answers is not in the document]
Comment (bmerry) [reply doc/fgpu.design.rst:350]: You've added a paragraph break, but you haven't introduced the fact that you're using example values - numbers like 4 and 32768 just appear out of nowhere.

File: doc/fgpu.design.rst [hunk 17]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+Now let's walk through each part. To recap the indexing used in the Cooley-Tukey algorithm: let a
+time-domain index :math:`i` be written as :math:`qn + r` and a frequency-domain index :math:`k` be
+written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r}, \dots, z_{(m-1)n+r}`,
+and denote its Fourier transform by :math:`Z^r`. It is worthwhile to point out that the superscript
+:math:`r` *does not* denote raise to the power but rather is a means to indicate an :math:`r^{th}` array.
+In practice this :math:`r^{th}` array is a subset (part) of the larger :math:`z` array of input data.
+
+Let's unpack this a bit further --- what is actually happening
+is that the initial array :math:`z` is divided into :math:`n=4` separate arrays each of
+:math:`m=32768/4 = 8192` elements (hence the :math:`N = mn` above)(This is true for a desired overall
```
Comment (bmerry) [inline doc/fgpu.design.rst:351]: Two parenthetical remarks next to each other looks odd and is hard to read.

File: doc/fgpu.design.rst [hunk 18]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+transform size of 32768 channels). The actual samples that land up in each array are defined by the
+indices :math:`i` and :math:`k`. Lets start with :math:`i`.
```
Comment (bmerry) [inline doc/fgpu.design.rst:353]:
```suggestion
indices :math:`i` and :math:`k`. Let's start with :math:`i`.
```

Comment (bmerry) [reply doc/fgpu.design.rst:353]: Ending a paragraph "Let's start with i" is a bit of a cliff-hanger. It might be more sensible to use that as the start of the next paragraph.

File: doc/fgpu.design.rst [hunk 19]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+========================   ========
+   Inputs                  Index(i)
+------------------------   --------
+  qn + r                      i
+========================   ========
+:math:`0 \cdot 4` + 0         0
+:math:`1 \cdot 4` + 0         4
+:math:`2 \cdot 4` + 0         8
+...               ...
+...               ...
```
Comment (bmerry) [inline doc/fgpu.design.rst:372]: Are the second set of dots in each line meant to be in the second column?

File: doc/fgpu.design.rst [hunk 20]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+range :math:`0` to :math:`m-1` (i.e. :math:`q=0` to :math:`q=8191`).
+So we are dividing up array :math:`z` into :math:`n` smaller arrays denoted by  :math:`r` (i.e. :math:`z^{r}`)
+each of length :math:`m=8192`. So what does this look like practically?
+
+The first array when :math:`r=0` (i.e. :math:`z^{0}`)
+
+========================   ========
+   Inputs                  Index(i)
+------------------------   --------
+  qn + r                      i
```
Comment (bmerry) [inline doc/fgpu.design.rst:366]: Not sure what it means to have two headings on a column, one of which is "Index(i)" and the other is "i".

File: doc/fgpu.design.rst [hunk 21]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+------------------------  --------
+  qn + r                     i
+========================  ========
+:math:`0 \cdot 4 + 3`        3
+:math:`1 \cdot 4 + 3`        7
+:math:`2 \cdot 4 + 3`        11
+...               ...
+...               ...
+:math:`8191 \cdot 4 + 3`   32767
+========================  ========
```
Comment (bmerry) [inline doc/fgpu.design.rst:391]: These tables take up an awful lot of space, particularly since you have a diagram as well. Can one just say something like, "For example, z^0 is z_0, z_4, z_8, ..., z_32764, while z^3 is z_3, z_7, z_11, ..., z_32767"? Or even just delete it and let the diagram speak for itself?

File: doc/fgpu.design.rst (hunk 10 above)
Comment (bmerry) [reply doc/fgpu.design.rst:397]: The image still references x rather than z; and the caption in the middle makes it sound like z^r are the outputs of a Fourier transform, rather than the inputs.

File: doc/fgpu.design.rst [hunk 22]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+What this shows is that each sub-array consists of samples from the initial array :math:`z` indexed
+by :math:`i=qn+r` where each sample is every :math:`4^{th}` and offset by :math:`r`. Pictorially
+this looks like,
+
+.. image:: images/z_array.png
+   :width: 600
+(click image to enlarge)
+
+Right, so we have separate sub-arrays as indexed from the initial array, what happens next? These various
+:math:`z^{r}` arrays are fed to cuFFT yielding 4 complex-to-complex transforms. These separate transforms
```
Comment (bmerry) [inline doc/fgpu.design.rst:402]:
```suggestion
:math:`z^{r}` arrays are fed to cuFFT yielding :math:`n` complex-to-complex transforms. These separate transforms
```

File: doc/fgpu.design.rst [hunk 23]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+now need to be combined to form a single real-to-complex transform of the full initial size. An inconvenience
+of this structure is that :math:`z^r` is not a contiguous set of input samples, but a strided array.
+While cuFFT does support both strided inputs and batched transformations, we cannot batch over :math:`r`
+and over multiple spectra at the same time as it only supports a single batch dimension with corresponding
+stride. We solve this by modifying the PFB kernel to reorder its output such that each :math:`z^r` is output
+contiguously. This can be done by shuffling some bits in the output index (because we assume powers of two
+everywhere).
+
+To see how the :math:`k` indexing works, :math:`k = pm + s` and is dealt with in a similar manner as above.
```
Comment (bmerry) [inline doc/fgpu.design.rst:411]: Comma splice.

File: doc/fgpu.design.rst [hunk 24]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+===========================  =========
+   Inputs                     Index(k)
+---------------------------  ---------
+  pm + s                      k
+===========================  =========
+:math:`0 \cdot 8192` + 0      0
+:math:`0 \cdot 8192` + 1      1
+:math:`0 \cdot 8192` + 2      2
+...               ...
+...               ...
```
Comment (bmerry) [inline doc/fgpu.design.rst:429]: Again, second set of ... seems to be in the wrong column.

File: doc/fgpu.design.rst [hunk 25]
```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
...
+---------------------------  --------
+  pm + s                         k
+===========================  ========
+:math:`3 \cdot 8192` + 0      24576
+:math:`3 \cdot 8192` + 1      24577
+:math:`3 \cdot 8192` + 2      24578
+...              ...
+...              ...
+:math:`3 \cdot 8192` + 8191   32767
+===========================  ========
```
Comment (bmerry) [inline doc/fgpu.design.rst:448]: Again, these tables take a huge amount of space and could be replaced by a couple of sentences.

**Review by bmerry — CHANGES_REQUESTED** [review]

**Comment by bmerry** [comment]
@james-smith-za adding you as a reviewer because I've tagged you on a couple of questions.

File: doc/fgpu.design.rst (hunk 13 above)
Comment (andrewvanderbyl) [reply doc/fgpu.design.rst:246]: The definition follows immediately after so is it a major issue?

File: doc/fgpu.design.rst (hunk 10 above)
Comment (andrewvanderbyl) [reply doc/fgpu.design.rst:397]: Previous feedback suggested the image be modified to x_0 to x_65535. I'll commit the source (.ppt).

Previous feedback:
The top line refers to z, but it should refer to x, and the values should be labelled x_0 to x_65535, not r_0 to r_65536.
It shows the z^r arrays each containing either only real or only imaginary components. These are complex arrays. At this stage we're not worrying about the real-to-complex transform, and I think the diagram would be less confusing if it only dealt with z and showed complex numbers without trying to split them into real and imaginary.

Comment (andrewvanderbyl) [reply doc/fgpu.design.rst:375]: Image updated as requested.

File: doc/fgpu.design.rst (hunk 13 above)
Comment (andrewvanderbyl) [reply doc/fgpu.design.rst:246]: Reordered to aid the flow.

**Comment by andrewvanderbyl** [comment]
Changes made as requested.

File: doc/fgpu.design.rst [hunk 26]
```diff
@@ -320,39 +312,23 @@ algorithm allows a transform of size :math:`N = mn` to be decomposed into :math:
...
 
-Let's unpack this a bit further --- what is actually happening
-is that the initial array :math:`z` is divided into :math:`n=4` separate arrays each of
-:math:`m=32768/4 = 8192` elements (hence the :math:`N = mn` above)(This is true for a desired overall
-transform size of 32768 channels). The actual samples that land up in each array are defined by the
-indices :math:`i` and :math:`k`. Lets start with :math:`i`.
+As a way of an example, let :math:`n=4` ("unzipping factor") and :math:`N=32768` (total number of channels).
+Now let's unpack this a bit further --- what is actually happening is that the initial array :math:`z` is divided into
+:math:`n=4` separate arrays each of :math:`m=32768/4 = 8192` elements (hence the :math:`N = mn` above).
+This is true for a desired overall transform size of 32768 channels. The actual samples that land up in each
```
Comment (bmerry) [inline doc/fgpu.design.rst:328]:
```suggestion
The actual samples that land up in each
```
You've already said above that the total number of channels is 32768.

File: doc/fgpu.design.rst [hunk 27]
```diff
@@ -361,45 +337,30 @@ each of length :math:`m=8192`. So what does this look like practically?
...
-:math:`8191 \cdot 4 + 3`   32767
-========================  ========
+This can be extended to the other remaining arrays. The fourth array when :math:`r=3` (for example), :math:`z^{3}`
+is :math:`z_{3}`, :math:`z_{7}`, :math:`z_{11}`, ..., :math:`z_{32767}`.
 
 What this shows is that each sub-array consists of samples from the initial array :math:`z` indexed
 by :math:`i=qn+r` where each sample is every :math:`4^{th}` and offset by :math:`r`. Pictorially
 this looks like,
 
 .. image:: images/z_array.png
```
Comment (bmerry) [inline doc/fgpu.design.rst:359]: The image still says "real-valued", should say "complex-valued".

**Review by bmerry — CHANGES_REQUESTED** [review]

File: doc/fgpu.design.rst (hunk 1 above)
Comment (james-smith-za) [reply doc/fgpu.design.rst:233]: I see your point, but I'm not a person who remembers the finer points of mathematical proofs from undergraduate days very well, so I feel as though it does add some value.

Does Sphinx do footnotes? That may be a good compromise.

**Review by bmerry — CHANGES_REQUESTED** [review]
For some reason Github is showing that I owe a review, even though I did a review since the last commit, so I'm hitting the "request changes" button again.

**Comment by andrewvanderbyl** [comment]
Changes made as per feedback by @bmerry and @james-smith-za.

**Review by james-smith-za — COMMENTED** [review]
Looks OK to me.

File: doc/fgpu.design.rst [hunk 28]
```diff
@@ -224,18 +224,9 @@ modulo :math:`N`.
...
-:math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the
-frequency component under computation for the input sequence :math:`x_i`. If :math:`k` is reversed
-(i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2n}\cdot i(k)}` as the negative
-in :math:`-k` multiplies out.
-
-Moving back to :math:`U` and :math:`V` components then :math:`U' = \overline{U}` and similarly :math:`V' = \overline{V}`.
-Going full circle (for completeness) if one was to conjugate the reversal (i.e. :math:`\overline{U'}`) then since
-:math:`U' = \overline{U}`, :math:`\overline{U'} = \overline{\overline{U}} = U`. Likewise for
-:math:`\overline{V'} = \overline{\overline{V}} = V`.
+Looking back at to :math:`U` and :math:`V` components then :math:`U' = \overline{U}` and similarly :math:`V' = \overline{V}`.
```
Comment (bmerry) [inline doc/fgpu.design.rst:229]: Something wrong with that grammar ("Looking back at to").

File: doc/fgpu.design.rst [hunk 29]
```diff
@@ -224,18 +224,9 @@ modulo :math:`N`.
 
 Hermitian symmetry means that :math:`U'_k = U_{-k} = \overline{U_k}` where
 the 'overline' in :math:`\overline{U_k}` denotes conjugation. This is effectively saying that by taking the
-reverse indices in :math:`U_k` we get a conjugated result.
+reverse indices in :math:`U_k` we get a conjugated result (see [#]_ for a reminder why this is the case).
```
Comment (bmerry) [inline doc/fgpu.design.rst:227]:
```suggestion
reverse indices in :math:`U_k` we get a conjugated result (see [#]_ for a reminder of why this is the case).
```

File: doc/fgpu.design.rst [hunk 30]
```diff
@@ -734,3 +724,7 @@ The work group size, subgroup size and coarsening factor can all affect
 performance significantly, and not always in obvious ways. It will likely be
 necessary to implement autotuning to get optimal results across a range of
 problem parameters and hardware devices, but this has not yet been done.
+
+.. rubric:: Footnotes
+
+.. [#] Going back to the original definition for the DFT we saw the complex exponential :math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the frequency component under computation for the input sequence :math:`x_i`. If :math:`k` is reversed (i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2n}\cdot i(k)}` as the negative in :math:`-k` multiplies out.
```
Comment (bmerry) [inline doc/fgpu.design.rst:730]: Wrap this rather long line - it will make future changes easier to review.

Comment (bmerry) [inline doc/fgpu.design.rst:730]:
```suggestion
.. [#] Going back to the original definition for the DFT we saw the complex exponential :math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the frequency component under computation for the input sequence :math:`x_i.` If :math:`k` is reversed (i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2n}\cdot i(k)}` as the negative in :math:`-k` multiplies out.
```

At the moment the full stop is getting wrapped onto the next line in the HTML output. Putting it inside the math prevents the line being broken in the wrong place.

Comment (bmerry) [inline doc/fgpu.design.rst:730]:
```suggestion
.. [#] Going back to the original definition for the DFT we saw the complex exponential :math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the frequency component under computation for the input sequence :math:`x_i`. If :math:`k` is reversed (i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2N}\cdot i(k)}` as the negative in :math:`-k` multiplies out.
```

**Review by bmerry — CHANGES_REQUESTED** [review]

File: doc/fgpu.design.rst (hunk 30 above)
Comment (bmerry) [reply doc/fgpu.design.rst:730]: Also it looks a bit odd that this line uses `ik` in the first formula for `i(k)` in the second.

**Review by bmerry — APPROVED** [review]
